utils/controller_debug.py

```diff
@@ -0,0 +1,789 @@
+
+    def cmdloop(self):
+        while True:
+
+            try:
+                return cmd.Cmd.cmdloop(self)
+
+            except Error as e:
+                print(e)
+
+            except:
+                traceback.print_exc()
+                # print sys.exc_info()
+                pass
+
+    def emptyline(self):
+        ReSync()
+        self.UpdatePrompt()
+        self.GetVarInfo()
+
+    def do_debug(self, line):
+        """Toggles display of low level serial data on/off"""
+        global showSerial
+        showSerial = not showSerial
+
+    def do_run(self, line):
+        "Run a Python script located in the scripts directory"
+        p = str.split(line)
+        if os.path.exists(p[0]):
+            fname = p[0]
+        elif os.path.exists("scripts/" + p[0]):
+            fname = "scripts/" + p[0]
+        else:
+            print("Unknown file " + p[0])
+            return
+        gbl = globals().copy()
+        gbl["cmdline"] = line
+        gbl["parser"] = self
+        execfile(fname, gbl)
+
+    def complete_run(self, text, line, begidx, endidx):
+        return glob.glob(text + "*.py") + [
+            x[8:] for x in glob.glob("scripts/" + text + "*.py")
+        ]
+
+    def do_exec(self, line):
+        exec(line)
+
+    def do_peek(self, line):
+        """
+Peek at a memory location.
+
+ex: peek <addr> <ct> <fmt> <file>
+
+   addr - the starting address
+   ct   - Number of bytes to read (default 1)
+   fmt  - An optional formatting string.
+   file - An optional file to save the data to
+
+   The formatting string determines how the data is interpreted and displayed.
+   Its a string made up of the following characters:
+     +  current address
+     x  16-bit integer displayed in hex
+     i  16-bit signed integer displayed in decimal
+     u  16-bit unsigned integer displayed in decimal
+     X  32-bit integer displayed in hex
+     I  32-bit signed integer displayed in decimal
+     U  32-bit unsigned integer displayed in decimal
+     f  32-bit float
+     e  32-bit float in exponential format
+     c  Single byte displayed as an ASCII character
+     b  Single byte displayed in hex
+
+   The data is extracted from what's returned and formatted as described in the string.
+   If there's more data left over at the end of the string, a new line starts and the
+   string starts over.
+
+   The default formatting string if none is supplied is +XXXX
+   i.e. Data is displayed as a series of 4 32-bit hex values / line
+"""
+        param = str.split(line)
+        addr = 0
+        ct = 1
+        fmt = "+XXXX"
+        fname = None
+        if len(param) > 3:
+            fname = param[3]
+        if len(param) > 2:
+            fmt = param[2]
+        if len(param) > 1:
+            ct = int(param[1], 0)
+        if len(param) > 0:
+            addr = param[0]
+        peek(addr, ct, fmt, fname)
+
+    def do_poke(self, line):
+        """
+Write data to a memory address
+
+ex: poke [type] <addr> <data>
```

Comment:
Ya, I started to make it more explicit but it was getting unreadable.  I just mentioned that the parameters are optional in their description.  Otherwise it ended up being something like:
peek <addr> [ ct [ fmt [ file ]]]